File: Arrays & Hashing/7. Valid Sudoku/solution.exs
defmodule Solution do
  @spec is_valid_sudoku(board :: [[char]]) :: boolean
  def is_valid_sudoku(board) do
    cols = Enum.zip_with(board, &List.flatten(&1))

    sub_box =
      board
      |> Enum.map(&Enum.chunk_every(&1, 3))
      |> Enum.zip_with(& &1)
      |> Enum.flat_map(&Enum.chunk_every(&1, 3))
      |> Enum.map(&Enum.reduce(&1, [], fn x, acc -> acc ++ x end))

    check(board) && check(cols) && check(sub_box)
  end

  def check([]), do: true

  def check([h | tail]) do
    letters = Enum.reject(h, & &1 == ?.)

    case Enum.uniq(letters) == letters do
      true -> check(tail)
      false -> false
    end
  end
end

File: Arrays & Hashing/7. Valid Sudoku/solution.go
func isValidSudoku(board [][]byte) bool {
	tmp := make(map[int]int)
	for i := 0; i < 9; i++ {
		for j := 0; j < 9; j++ {
			n := board[i][j]
			if n != '.' {
				tmp[int(n)*10+i]++
				tmp[int(n)*100+j]++
				tmp[int(n)*1000+(i/3*10+j/3)]++
				if tmp[int(n)*10+i] > 1 || tmp[int(n)*100+j] > 1 || tmp[int(n)*1000+(i/3*10+j/3)] > 1 {
					return false
				}
			}
		}
	}
	return true
}
